panic: runtime error by running `gh collaborators list`

Hello!

I ran `gh collaborators list ${GITHUB_ORG}` and I got the following panic error: (GITHUB_ORG redacted).

```
2025/10/24 12:30:14 Body read error, HTTP 403: You must be an owner of this organization to list outside collaborators. (https://api.github.com/orgs/{GITHUB_ORG}/outside_collaborators)
panic: runtime error: invalid memory address or nil pointer dereference
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x40 pc=0x1022fcf0c]

goroutine 1 [running]:
github.com/katiem0/gh-collaborators/internal/utils.(*APIGetter).GetOrgGuestCollaborators.func1()
	github.com/katiem0/gh-collaborators/internal/utils/general.go:43 +0x1c
panic({0x1023f0be0?, 0x102672330?})
	runtime/panic.go:785 +0x124
github.com/katiem0/gh-collaborators/internal/utils.(*APIGetter).GetOrgGuestCollaborators(0x14000147ba0, {0x16dde7156?, 0x102312e62?})
	github.com/katiem0/gh-collaborators/internal/utils/general.go:48 +0x1ac
github.com/katiem0/gh-collaborators/cmd/list.runCmdList({0x16dde7156, 0x13}, 0x1400003e280, 0x14000279ba0, {0x102456f68, 0x14000200038?})
	github.com/katiem0/gh-collaborators/cmd/list/list.go:131 +0x304
github.com/katiem0/gh-collaborators/cmd/list.NewCmdList.func1(0x1400015a400?, {0x140000324c0, 0x1, 0x102301ff7?})
	github.com/katiem0/gh-collaborators/cmd/list/list.go:96 +0x46c
github.com/spf13/cobra.(*Command).execute(0x1400015c908, {0x14000032490, 0x1, 0x1})
	github.com/spf13/cobra@v1.9.1/command.go:1015 +0x82c
github.com/spf13/cobra.(*Command).ExecuteC(0x1400015c308)
	github.com/spf13/cobra@v1.9.1/command.go:1148 +0x350
github.com/spf13/cobra.(*Command).Execute(0x1026414d8?)
	github.com/spf13/cobra@v1.9.1/command.go:1071 +0x1c
main.main()
	github.com/katiem0/gh-collaborators/main.go:12 +0x20
```

It would be expected to only get the API HTTP 403 error and quit. This seems to be related to https://github.com/katiem0/gh-collaborators/issues/9

--- a/cmd/list/list.go
+++ b/cmd/list/list.go
@@ -3,9 +3,7 @@ package list
 import (
 	"encoding/csv"
 	"encoding/json"
-	"errors"
 	"fmt"
-	"io"
 	"os"
 	"strconv"
 	"time"
@@ -83,24 +81,25 @@ func NewCmdList() *cobra.Command {
 
 			owner := args[0]
 
-			if _, err := os.Stat(cmdFlags.listFile); errors.Is(err, os.ErrExist) {
-				return err
+			// Check if file exists, but don't fail if it doesn't
+			if _, err := os.Stat(cmdFlags.listFile); err == nil {
+				return fmt.Errorf("output file %s already exists", cmdFlags.listFile)
 			}
 
-			reportWriter, err := os.OpenFile(cmdFlags.listFile, os.O_WRONLY|os.O_CREATE, 0644)
+			// Create APIGetter
+			apiGetter := utils.NewAPIGetter(gqlClient, restClient)
 
-			if err != nil {
+			// Collect all data first, don't create file yet
+			if err := runCmdList(owner, &cmdFlags, apiGetter); err != nil {
 				return err
 			}
-
-			return runCmdList(owner, &cmdFlags, utils.NewAPIGetter(gqlClient, restClient), reportWriter)
+			return nil
 		},
 	}
 
 	reportFileDefault := fmt.Sprintf("RepoCollaboratorsReport-%s.csv", time.Now().Format("20060102150405"))
 
 	// Configure flags for command
-
 	listCmd.PersistentFlags().StringVarP(&cmdFlags.token, "token", "t", "", `GitHub Personal Access Token (default "gh auth token")`)
 	listCmd.PersistentFlags().StringVarP(&cmdFlags.hostname, "hostname", "", "github.com", "GitHub Enterprise Server hostname")
 	listCmd.Flags().StringVarP(&cmdFlags.listFile, "output-file", "o", reportFileDefault, "Name of file to write CSV list to")
@@ -110,12 +109,12 @@ func NewCmdList() *cobra.Command {
 	return listCmd
 }
 
-func runCmdList(owner string, cmdFlags *cmdFlags, g *utils.APIGetter, reportWriter io.Writer) error {
+func runCmdList(owner string, cmdFlags *cmdFlags, g *utils.APIGetter) error {
 	var reposCursor *string
+	var csvData [][]string
 
-	csvWriter := csv.NewWriter(reportWriter)
-
-	err := csvWriter.Write([]string{
+	// Add header to data slice
+	csvData = append(csvData, []string{
 		"RepositoryName",
 		"RepositoryID",
 		"Visibility",
@@ -123,34 +122,33 @@ func runCmdList(owner string, cmdFlags *cmdFlags, g *utils.APIGetter, reportWrit
 		"AccessLevel",
 	})
 
-	if err != nil {
-		zap.S().Error("Error raised in writing output", zap.Error(err))
-	}
-
 	zap.S().Debugf("Gathering repositories and access for %s", owner)
 	repoCollabList, err := g.GetOrgGuestCollaborators(owner)
 	if err != nil {
-		zap.S().Error("Error raised in gathering users", zap.Error(err))
+		zap.S().Errorf("Failed to get organization collaborators for '%s'", owner)
+		return err
 	}
 
 	var repoCollaborators []data.RepoCollaborators
 	err = json.Unmarshal(repoCollabList, &repoCollaborators)
 	if err != nil {
-		return err
+		zap.S().Errorf("Failed to parse collaborators data: %v", err)
+		return fmt.Errorf("failed to parse collaborators data: %w", err)
 	}
 
 	if len(cmdFlags.username) > 0 {
 		zap.S().Debugf("Checking if username %s is in list of repository collaborators", cmdFlags.username)
 		for _, repoCollab := range repoCollaborators {
 			if cmdFlags.username == repoCollab.Login {
-
 				zap.S().Debugf("Gathering repositories for specified username %s", cmdFlags.username)
 				var allRepoPerms []data.RepoInfo
 				for {
 					repoUserPermissions, err := g.GetOrgRepositoryPermissions(owner, cmdFlags.username, reposCursor)
 					if err != nil {
-						zap.S().Error("Error raised in gathering repositories and user permissions", zap.Error(err))
+						zap.S().Errorf("Failed to get repository permissions for user '%s' in organization '%s': %v", cmdFlags.username, owner, err)
+						return fmt.Errorf("failed to get repository permissions for user %s: %w", cmdFlags.username, err)
 					}
+
 					allRepoPerms = append(allRepoPerms, repoUserPermissions.Organization.Repositories.Nodes...)
 					if !repoUserPermissions.Organization.Repositories.PageInfo.HasNextPage {
 						break
@@ -159,21 +157,17 @@ func runCmdList(owner string, cmdFlags *cmdFlags, g *utils.APIGetter, reportWrit
 				}
 				for _, repo := range allRepoPerms {
 					if len(repo.Collaborators.Edges) > 0 {
-						err = csvWriter.Write([]string{
+						csvData = append(csvData, []string{
 							repo.Name,
 							strconv.Itoa(repo.DatabaseId),
 							repo.Visibility,
 							cmdFlags.username,
 							repo.Collaborators.Edges[0].Permission,
 						})
-						if err != nil {
-							zap.S().Error("Error raised in writing output", zap.Error(err))
-						}
 					}
 				}
 			}
 		}
-
 	} else {
 		for _, repoCollab := range repoCollaborators {
 			zap.S().Debugf("Gathering repositories for username %s", repoCollab.Login)
@@ -181,8 +175,10 @@ func runCmdList(owner string, cmdFlags *cmdFlags, g *utils.APIGetter, reportWrit
 			for {
 				repoUserPermissions, err := g.GetOrgRepositoryPermissions(owner, repoCollab.Login, reposCursor)
 				if err != nil {
-					zap.S().Error("Error raised in gathering repositories and user permissions", zap.Error(err))
+					zap.S().Errorf("Failed to get repository permissions for user '%s' in organization '%s': %v", repoCollab.Login, owner, err)
+					return fmt.Errorf("failed to get repository permissions for user %s: %w", repoCollab.Login, err)
 				}
+
 				allRepoPerms = append(allRepoPerms, repoUserPermissions.Organization.Repositories.Nodes...)
 				if !repoUserPermissions.Organization.Repositories.PageInfo.HasNextPage {
 					break
@@ -191,23 +187,49 @@ func runCmdList(owner string, cmdFlags *cmdFlags, g *utils.APIGetter, reportWrit
 			}
 			for _, repo := range allRepoPerms {
 				if len(repo.Collaborators.Edges) > 0 {
-					err = csvWriter.Write([]string{
+					csvData = append(csvData, []string{
 						repo.Name,
 						strconv.Itoa(repo.DatabaseId),
 						repo.Visibility,
 						repoCollab.Login,
 						repo.Collaborators.Edges[0].Permission,
 					})
-					if err != nil {
-						zap.S().Error("Error raised in writing output", zap.Error(err))
-					}
 				}
 			}
 		}
 	}
 
-	fmt.Printf("Successfully listed repository collaborator permissions for repositories in %s", owner)
-	csvWriter.Flush()
+	// Only create and write to file after all data is successfully collected
+	if len(csvData) <= 1 { // Only header, no actual data
+		return fmt.Errorf("no collaborator data found for organization %s", owner)
+	}
+
+	zap.S().Debugf("Creating output file %s", cmdFlags.listFile)
+	reportWriter, err := os.OpenFile(cmdFlags.listFile, os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		return fmt.Errorf("failed to create output file: %w", err)
+	}
+	defer func() {
+		closeErr := reportWriter.Close()
+		if closeErr != nil {
+			zap.S().Warnf("Error closing file: %v", closeErr)
+		}
+	}()
+
+	csvWriter := csv.NewWriter(reportWriter)
+	defer csvWriter.Flush()
+
+	// Write all collected data to CSV
+	for _, row := range csvData {
+		err = csvWriter.Write(row)
+		if err != nil {
+			zap.S().Error("Error raised in writing output", zap.Error(err))
+			return fmt.Errorf("failed to write CSV data: %w", err)
+		}
+	}
+
+	fmt.Printf("Successfully listed repository collaborator permissions for repositories in %s\n", owner)
+	fmt.Printf("Report saved to: %s\n", cmdFlags.listFile)
 
 	return nil
 }

--- a/internal/log/log.go
+++ b/internal/log/log.go
@@ -5,7 +5,6 @@ import (
 )
 
 func NewLogger(debug bool) (*zap.Logger, error) {
-
 	level := zap.InfoLevel
 
 	if debug {
@@ -18,6 +17,15 @@ func NewLogger(debug bool) (*zap.Logger, error) {
 		EncoderConfig:    zap.NewDevelopmentEncoderConfig(),
 		OutputPaths:      []string{"stderr"},
 		ErrorOutputPaths: []string{"stderr"},
+		// Disable stack traces for cleaner output
+		DisableStacktrace: !debug, // Only show stack traces in debug mode
+	}
+
+	// Customize encoder config for better readability
+	if debug {
+		loggerConfig.EncoderConfig.StacktraceKey = "stacktrace"
+	} else {
+		loggerConfig.EncoderConfig.StacktraceKey = ""
 	}
 
 	return loggerConfig.Build()

--- a/internal/utils/general.go
+++ b/internal/utils/general.go
@@ -4,6 +4,7 @@ import (
 	"fmt"
 	"io"
 	"log"
+	"strings"
 
 	"github.com/cli/go-gh/v2/pkg/api"
 	"github.com/katiem0/gh-collaborators/internal/data"
@@ -44,19 +45,29 @@ func (g *APIGetter) GetOrgGuestCollaborators(owner string) ([]byte, error) {
 	zap.S().Debugf("Reading in repository collaborators from %v", url)
 	resp, err := g.restClient.Request("GET", url, nil)
 	if err != nil {
-		log.Printf("Body read error, %v", err)
-	}
-	defer func() {
-		closeErr := resp.Body.Close()
-		if closeErr != nil {
-			zap.S().Warnf("Error closing response body: %v", closeErr)
+		// Check for specific permission error
+		if strings.Contains(err.Error(), "403") && strings.Contains(err.Error(), "must be an owner") {
+			return nil, fmt.Errorf("insufficient permissions: you must be an owner of the organization '%s' to list outside collaborators", owner)
 		}
-	}()
+		zap.S().Errorf("Error making request to %s: %v", url, err)
+		return nil, err
+	}
+
+	if resp != nil && resp.Body != nil {
+		defer func() {
+			closeErr := resp.Body.Close()
+			if closeErr != nil {
+				zap.S().Warnf("Error closing response body: %v", closeErr)
+			}
+		}()
+	}
+
 	responseData, err := io.ReadAll(resp.Body)
 	if err != nil {
-		log.Printf("Body read error, %v", err)
+		zap.S().Errorf("Body read error: %v", err)
+		return nil, err
 	}
-	return responseData, err
+	return responseData, nil
 }
 
 func (g *APIGetter) GetOrgRepositoryPermissions(owner string, user string, endCursor *string) (*data.OrganizationUserQuery, error) {

